### ISSUE TITLE
[BUG] Loading Artifacts failed in UI when using azureml backend store

### ISSUE THREAD

**Issue by Gabriel2409** [mlflow/mlflow#9495]
### Willingness to contribute

Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.

### MLflow version

2.6.0

### System information

- **OS Platform and Distribution**:  Linux Ubuntu 22.04
- **Python version**: 3.10.11

### Describe the problem

Problem: 
When using the mlflow ui with an azureml backend store, the artifacts are not loaded even though the other metrics are showing as intended. 

![image](https://github.com/mlflow/mlflow/assets/12717469/4be04398-655a-48bc-9fbf-7fa6308ccf77)

In the terminal, I have the following error: 

```
Traceback (most recent call last):
  File "/home/gcournelle/workprojects/mlops_template/venv/lib/python3.10/site-packages/flask/app.py", line 2190, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/gcournelle/workprojects/mlops_template/venv/lib/python3.10/site-packages/flask/app.py", line 1486, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/gcournelle/workprojects/mlops_template/venv/lib/python3.10/site-packages/flask/app.py", line 1484, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/gcournelle/workprojects/mlops_template/venv/lib/python3.10/site-packages/flask/app.py", line 1469, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/home/gcournelle/workprojects/mlops_template/venv/lib/python3.10/site-packages/mlflow/server/handlers.py", line 476, in wrapper
    return func(*args, **kwargs)
  File "/home/gcournelle/workprojects/mlops_template/venv/lib/python3.10/site-packages/mlflow/server/handlers.py", line 517, in wrapper
    return func(*args, **kwargs)
  File "/home/gcournelle/workprojects/mlops_template/venv/lib/python3.10/site-packages/mlflow/server/handlers.py", line 953, in _list_artifacts
    artifact_entities = _get_artifact_repo(run).list_artifacts(path)
AttributeError: 'Response' object has no attribute 'list_artifacts'
```

I suspect the error may come from the fact that the artifacts are not where mlflow expect them. 
Here is where they were stored in my blob storage: `https://<mystorage>.blob.core.windows.net/azureml/ExperimentRun/dcid.3bacddc1-79a5-42df-9968-f65b7e8a458d/`

### Steps to reproduce the bug

- Create an azure machine learning workspace
- Run the code below by replacing the following variables with the ones from your azure account: 
region, subscription_id, resource_group, workspace_name
- Launch the ui with : `mlflow ui --backend-store-uri azureml://<region>.api.azureml.ms/mlflow/v1.0/subscriptions/<subscription_id>/resourceGroups/<resource_group>/providers/Microsoft.MachineLearningServices/workspaces/<workspace_name>` (replace variables between <>)
- Navigate to localhost:5000 and select the experiment called `myexperiment`
- Then click on the last run
![image](https://github.com/mlflow/mlflow/assets/12717469/aeacc3ba-7b12-49ec-a06b-a4ef1c71611b)

### Code to generate data required to reproduce the bug

```python
import mlflow
import os
from random import random

region = "westeurope"
subscription_id = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
resource_group = "myrg"
workspace_name = "myws"

azureml_mlflow_uri = f"azureml://{region}.api.azureml.ms/mlflow/v1.0/subscriptions/{subscription_id}/resourceGroups/{resource_group}/providers/Microsoft.MachineLearningServices/workspaces/{workspace_name}"

mlflow.set_tracking_uri(azureml_mlflow_uri)

experiment_name = 'myexperiment'
mlflow.set_experiment(experiment_name)

from random import random
with mlflow.start_run() as mlflow_run:
    mlflow.log_param("hello_param", "world")
    mlflow.log_metric("hello_metric", random())
    os.system(f"echo 'hello world' > helloworld.txt")
    mlflow.log_artifact("helloworld.txt")

experiment_id = mlflow.get_experiment_by_name(experiment_name).experiment_id
run_id = mlflow.search_runs([experiment_id]).loc[0, 'run_id']
mlflow.artifacts.download_artifacts(run_id=run_id) # here, it works as intended

# Issue comes when we are trying to see artifacts in the ui 
````

### Is the console panel in DevTools showing errors relevant to the bug?

![image](https://github.com/mlflow/mlflow/assets/12717469/4b37d723-3a91-482e-94aa-dccd3ec5992e)

```
Fetch failed:  
Response { type: "basic", url: "http://127.0.0.1:5000/ajax-api/2.0/mlflow/artifacts/list?run_uuid=3bacddc1-79a5-42df-9968-f65b7e8a458d", redirected: false, status: 500, ok: false, statusText: "INTERNAL SERVER ERROR", headers: Headers(5), body: ReadableStream, bodyUsed: true }
body: ReadableStream { locked: true }
bodyUsed: true
headers: Headers(5) { connection → "close", "content-length" → "265", "content-type" → "text/html; charset=utf-8", … }
ok: false
redirected: false
status: 500
statusText: "INTERNAL SERVER ERROR"
type: "basic"
url: "http://127.0.0.1:5000/ajax-api/2.0/mlflow/artifacts/list?run_uuid=3bacddc1-79a5-42df-9968-f65b7e8a458d"
<prototype>: ResponsePrototype { clone: clone(), arrayBuffer: arrayBuffer(), blob: blob(), … }
FetchUtils.ts:90:10
```

### Does the network panel in DevTools contain failed requests relevant to the bug?

![image](https://github.com/mlflow/mlflow/assets/12717469/b8669a9d-3d83-439f-a59a-bc0ad54ad52e)
```
GET /ajax-api/2.0/mlflow/artifacts/list?run_uuid=3bacddc1-79a5-42df-9968-f65b7e8a458d HTTP/1.1
Host: 127.0.0.1:5000
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/117.0
Accept: */*
Accept-Language: fr,fr-FR;q=0.8,en-US;q=0.5,en;q=0.3
Accept-Encoding: gzip, deflate, br
Referer: http://127.0.0.1:5000/
Content-Type: application/json; charset=utf-8
Connection: keep-alive
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-origin


HTTP/1.1 500 INTERNAL SERVER ERROR
Server: gunicorn
Date: Thu, 31 Aug 2023 09:12:55 GMT
Connection: close
Content-Type: text/html; charset=utf-8
Content-Length: 265
```

**Comment by serena-ruan** [comment]
Hi @akshaya-a Any clues or insights on this? Thanks!

**Comment by akshaya-a** [comment]
this has annoyed me as well! thanks @serena-ruan  you can assign it my way. @Gabriel2409 is this blocking you or a curiosity? Most AML customers end up using our ml.azure.com UI instead but I'm happy to dig into this if this is something useful for your use case - mostly looking to understand priority

**Comment by Gabriel2409** [comment]
Hi @akshaya-a, thank you for your reply. 
This is actually blocking for me as I need to use the standard mlflow ui. 

After further investigation, I think the error comes from here, in `mlflow.server.handlers.py`:


```python 
@catch_mlflow_exception
def _get_artifact_repo(run):
    return get_artifact_repository(run.info.artifact_uri)
```

- `run.info.artifact_uri` looks like this: `azureml://<region>.api.azureml.ms/mlflow/v2.0/subscriptions/<subscriptionid>/resourceGroups/<resourcegroup>/providers/Microsoft.MachineLearningServices/workspaces/<workspacename>/experiments/<experimentid>/runs/<runid>/artifacts/`
- I added a breakpoint here and ran the `get_artifact_repository` function and I got the following error: `*** mlflow.exceptions.MlflowException: An authorization token was not set in the artifact uri`

What is strange is that I am authenticated in azure.

**Comment by akshaya-a** [comment]
OK @Gabriel2409 workaround if you can: set MLFLOW_TRACKING_URI to the same value, not just --backend-store-uri. That should get you further (i can now list and fetch artifact contents in the UI for the few I spot checked).

It comes from deep in the bowels relying on MLflowClient() to do the right thing, which in regular jobs/scripts normally does due to the env vars being set. I'll need a little time to ponder how the other plugins correctly fetch the context from the CLI and rework our flow to do the same.

The error message is misleading because it's expecting to be in another path where there is a SAS URI, but we got there because MLflowClient() resolves to FileStore, not the AzureMLRestStore, hence the workaround.

**Comment by Gabriel2409** [comment]
@akshaya-a this works like a charm. I would not have imagined the fix would be so easy, you are amazing, thank you!

To tell you the full story, it was basically politics. My team already used mlflow without azureml and we transitioned to azureml recently. However, there was lot of push back because they did not want to use the azure ml UI to track experiments. I suspect this will change once they become used to it but for now having a workaround to use the standard UI is very helpful. 

Once again, thank you very much.

**Comment by akshaya-a** [comment]
I understand! Change is hard so making that transition easier is a worthy goal - for a different customer I tried a fun prototype to make using the mlflow UI a bit easier here, let me know if it helps: https://github.com/akshaya-a/azureml-apps

@serena-ruan  I'll keep this bug open until I fix the real issue but now that Gabriel is unblocked, could take a bit

**Comment by akshaya-a** [comment]
I think I found the real issue, and while I don't necessarily like our plugin's code, the fix I propose above would make any other plugin implementation less error-prone as well. It's not really related to our plugin so I created a new issue @serena-ruan, let me know your thoughts. A PR fixing that would then close this at the same time

**Comment by akshaya-a** [comment]
i think we can close this @BenWilson2 @harupy @serena-ruan
